src/evict.c

```diff
@@ -581,60 +624,61 @@ int freeMemoryIfNeeded(void) {
-                    /* Let's satisfy our stop condition. */
-                    mem_freed = mem_tofree;
+            if (keys_freed % 16 == 0) {
+                /* When the memory to free starts to be big enough, we may
+                 * start spending so much time here that is impossible to
+                 * deliver data to the replicas fast enough, so we force the
+                 * transmission here inside the loop. */
+                if (slaves) flushSlavesOutputBuffers();
+
+                /* Normally our stop condition is the ability to release
+                 * a fixed, pre-computed amount of memory. However when we
+                 * are deleting objects in another thread, it's better to
+                 * check, from time to time, if we already reached our target
+                 * memory, since the "mem_freed" amount is computed only
+                 * across the dbAsyncDelete() call, while the thread can
+                 * release the memory all the time. */
+                if (server.lazyfree_lazy_eviction) {
+                    if (getMaxmemoryState(NULL,NULL,NULL,NULL) == C_OK) {
+                        break;
+                    }
+                }
+
+                /* After some time, exit the loop early - even if memory limit
+                 * hasn't been reached.  If we suddenly need to free a lot of
+                 * memory, don't want to spend too much time here.  */
+                if (elapsedUs(evictionTimer) > EVICTION_PERIOD_US) {
+                    // We still need to free memory - start eviction timer proc
+                    if (!isEvictionProcRunning) {
+                        isEvictionProcRunning = 1;
+                        aeCreateTimeEvent(server.el, 0,
+                                evictionTimeProc, NULL, NULL);
+                    }
+                    break;
                 }
             }
         } else {
             goto cant_free; /* nothing to free... */
         }
     }
-    result = C_OK;
+    /* at this point, the memory is OK, or we have reached the time limit */
+    result = (isEvictionProcRunning) ? EVICT_RUNNING : EVICT_OK;
 
 cant_free:
-    /* We are here if we are not able to reclaim memory. There is only one
-     * last thing we can try: check if the lazyfree thread has jobs in queue
-     * and wait... */
-    if (result != C_OK) {
-        latencyStartMonitor(lazyfree_latency);
-        while(bioPendingJobsOfType(BIO_LAZY_FREE)) {
+    if (result == EVICT_FAIL) {
+        /* At this point, we have run out of evictable items.  It's possible
+         * that some items are being freed in the lazyfree thread.  Perform a
+         * short wait here if such jobs exist, but don't wait long.  */
+        if (bioPendingJobsOfType(BIO_LAZY_FREE)) {
+            usleep(EVICTION_PERIOD_US);
             if (getMaxmemoryState(NULL,NULL,NULL,NULL) == C_OK) {
-                result = C_OK;
-                break;
+                result = EVICT_OK;
```

Comment:
i understand why you remove the loop (latency), but in that case as long as there's something in the lazy free list, we should return with PENDING, not FAIL.
and if we do that, maybe there's no need for the usleep that's still here.

Reply:
My logic:
* If we've exited the main loop above with EVICT_FAIL, the main loop has exhausted all evictable items.  Though there are some evictable items still being processed by lazy free, we have no idea how many or if there are enough to overcome the current memory deficiency.  The safe thing to do in this situation is to return EVICT_FAIL.  (Even if we did return EVICT_PASS, it's almost certain that one of the next commands will fail anyway.  So it becomes a situation of trading one failure for another.)
* I had considered removing the sleep entirely, but decided that it was better to slightly slow the main thread - giving some time to the background thread to perform evictions.  This will perform better than the existing code and is a bit more conservative than removing the sleep altogether.  ALSO - if the sleep is removed, there's little point in calling `getMemoryState` and therefore no reason to check for `bioPendingJobs` (so the entire `if` clause can be removed).